deepfence_server/ingesters/agent.go

```diff
@@ -237,168 +238,139 @@ func concatMaps(input map[string][]string) []map[string]interface{} {
+	host_batch := make([]map[string]interface{}, 0, len(rpt.Host))
+	kubernetes_batch := make([]map[string]interface{}, 0, len(rpt.KubernetesCluster))
+	kubernetes_edges_batch := map[string][]string{}
+
+	for _, n := range rpt.Host {
+		hosts = append(hosts, map[string]interface{}{"node_id": n.NodeID})
+		host_batch = append(host_batch, utils.StructToMap(n))
+	}
+
+	for nodeId, parent := range rpt.HostParents {
+		if parent.KubernetesCluster != "" {
+			kubernetes_edges_batch[parent.KubernetesCluster] = append(kubernetes_edges_batch[parent.KubernetesCluster], nodeId)
 		}
 	}
 
+	for _, n := range rpt.KubernetesCluster {
+		kubernetes_batch = append(kubernetes_batch, utils.StructToMap(n))
+	}
+
 	processes_to_keep := map[string]struct{}{}
 	var buf bytes.Buffer
-	for _, n := range rpt.Endpoint.Nodes {
-		node_info := n.ToDataMap()
-		if hni, ok := node_info["host_node_id"]; ok {
-			host := extractHostFromHostNodeID(hni)
-			buf.Reset()
-			buf.WriteString(host)
-			buf.WriteByte(';')
-			buf.WriteString(node_info["pid"])
-			host_pid := buf.String()
-			processes_to_keep[host_pid] = struct{}{}
+	for _, n := range rpt.Endpoint {
+		if n.HostName == "" {
+			continue
 		}
+		buf.Reset()
+		buf.WriteString(n.HostName)
+		buf.WriteByte(';')
+		buf.WriteString(strconv.Itoa(n.Pid))
+		host_pid := buf.String()
+		processes_to_keep[host_pid] = struct{}{}
 	}
 
-	endpoint_edges_batch := make([]map[string]interface{}, 0, len(rpt.Endpoint.Nodes))
+	endpoint_edges_batch := make([]map[string]interface{}, 0, len(rpt.Endpoint))
 	//endpoint_batch := []map[string]string{}
 	//endpoint_edges := []map[string]string{}
-	for _, n := range rpt.Endpoint.Nodes {
-		node_info := n.ToDataMap()
-		if _, ok := node_info["host_node_id"]; !ok {
-			node_ip, _ := extractIPPortFromEndpointID(node_info["node_id"])
+	// Handle inbound from internet
+	inbound_edges := []map[string]string{}
+
+	for _, n := range rpt.Endpoint {
+		hostName := n.HostName
+		if hostName == "" {
+			node_ip, _ := extractIPPortFromEndpointID(n.NodeID)
 			if val, ok := resolvers.get_host(node_ip); ok {
-				node_info["host_node_id"] = val
+				hostName = val
 			} else {
 				// This includes skipping all endpoint having 127.0.0.1
 				continue
 			}
 		}
-		//node_info["adja"] = strings.Join(n.Adjacency, ",")
-		//endpoint_batch = append(endpoint_batch, node_info)
-
-		host_name := extractHostFromHostNodeID(node_info["host_node_id"])
-		if len(node_info["pid"]) > 0 {
-			edges := make([]map[string]string, 0, len(n.Adjacency))
-			for _, i := range n.Adjacency {
-				if n.ID != i {
-					ip, port := extractIPPortFromEndpointID(i)
-					if host, ok := resolvers.get_host(ip); ok {
-						if host_name == host {
-							continue
-						}
-						right_ippid, ok := resolvers.get_ip_pid(ip + port)
-						if ok {
-							rightpid := extractPidFromNodeID(right_ippid)
+		if n.Pid >= 0 {
+			adjacency, ok := rpt.EndpointAdjacency[n.NodeID]
+			if !ok {
+				continue
+			}
+			if len(adjacency) == 0 {
```

Comment:
Not sure we need this condition?
I don't see it in prev code

Reply:
There were 2 for loops, merged to one, so this condition. (mostly to avoid few lines of duplicate code)
https://github.com/deepfence/ThreatMapper/blob/v2/deepfence_server/ingesters/agent.go#L273-L340

Reply:
Ah I see - I personally like it separated because 1 code section is about linking "internal" connections, while the other code section is about linking "external" connections. But I get your point regarding code dup.
Can we just put back the comment: `// Handle inbound from internet` to indicate `len(adjacency) == 0` meaning more clearly